pkl-core/src/main/java/org/pkl/core/ast/expression/generator/GeneratorForNode.java

```diff
@@ -147,84 +129,55 @@ protected void fallback(VirtualFrame frame, Object parent, ObjectData data, Obje
         .build();
   }
 
-  @SuppressWarnings("ForLoopReplaceableByForEach")
   private void doEvalObject(VirtualFrame frame, VmObject iterable, Object parent, ObjectData data) {
-    initTypeNodes(frame);
-    var members = evaluateMembers(iterable);
-    for (int i = 0; i < members.size(); i++) {
-      var member = members.get(i);
-      executeIteration(frame, parent, data, member.first, member.second);
-    }
-    resetFrameSlots(frame);
+    iterable.forceAndIterateMemberValues(
+        (key, member, value) -> {
+          var convertedKey = member.isProp() ? key.toString() : key;
+          // TODO: Executing iteration behind a Truffle boundary is bad for performance.
+          // This and similar cases will be fixed in an upcoming PR that replaces method
+          // `(forceAnd)iterateMemberValues` with cursor-based external iterators.
```

Comment:
Definitely interested in a future PR of yours that addresses this.

But, we currently have a (small) optimization that this PR removes, in `evaluateMembers`. I think we should bring that back.

Reply:
There are several places in the codebase that call Node.execute from behind a Truffle boundary. The PR that fixes all of them has been ready from my side for over a month, but due to the slow review progress, I haven't sent it yet. Please let's not bring back this workaround; it's not worth it and will only slow us down further.